File: app/java/src/main/java/examples/Node.java
package examples;

import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.json.JSONObject;
import utils.TransferClient;
import utils.Configuration;
import utils.Rest;

public class Node {

    private static final Logger LOGGER = Logger.getLogger(Node.class.getName());

    public static void main(String... args) throws Exception {
        final Configuration config = new Configuration(args);
        final TransferClient transferClient = new TransferClient(config);

        try {
            /*
             * Create Node API object
             */
            LOGGER.log(Level.INFO, "Creating Node API object");
            final String nodeBaseUrl = config.getParamStr("node", "url");
            final var nodeAPI = new Rest(nodeBaseUrl);
            nodeAPI.setVerify(config.getParamBool("node", "verify"));
            nodeAPI.setAuthBasic(config.getParamStr("node", "username"),
                    config.getParamStr("node", "password"));

            /*
             * Generate transfer spec using upload_setup
             */
            LOGGER.log(Level.INFO, "Generating transfer spec");
            final JSONObject uploadSetupRequest = new JSONObject().put("transfer_requests",
                    new org.json.JSONArray().put(new JSONObject().put("transfer_request",
                            new JSONObject().put("paths",
                                    new org.json.JSONArray().put(new JSONObject().put("destination",
                                            config.getParamStr("node", "folder_upload")))))));

            final JSONObject response =
                    (JSONObject) nodeAPI.create("files/upload_setup", uploadSetupRequest);
            final JSONObject transferSpec = response.getJSONArray("transfer_specs").getJSONObject(0)
                    .getJSONObject("transfer_spec");

            /*
             * Add local file paths to transfer spec
             */
            config.addSources(transferSpec, "paths", null);

            /*
             * Start transfer
             */
            LOGGER.log(Level.INFO, "Starting transfer");
            transferClient.start_transfer_and_wait(transferSpec);

        } finally {
            transferClient.shutdown();
        }
    }
}

File: app/ruby/src/examples/node.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Example: transfer a file using one of the provided transfer agents
# location of ascp can be specified with env var "ascp"
# temp folder can be specified with env var "tmp"

require_relative '../utils/configuration'
require_relative '../utils/transfer_client'
require_relative '../utils/rest'

# Initialize configuration
config = Utils::Configuration.instance

# Initialize transfer client
transfer_client = Utils::TransferClient.new(config).startup

begin
  destination_folder = config.param('server', 'folder_upload')

  ##############################################################
  # Upload with node authorization
  # create rest client for Node API on a public demo system, using public demo credentials
  node_api = Utils::Rest.new(config.param('node', 'url'))
  node_api.auth_basic(
    config.param('node', 'username'),
    config.param('node', 'password')
  )
  # Request transfer authorization to node for a single transfer (This is a node api v3 call)
  send_result = node_api.create(
    'files/upload_setup',
    { transfer_requests: [{ transfer_request: { paths: [{ destination: destination_folder }] } }] }
  )
  # we normally have only one transfer spec in list, so just get the first transfer_spec
  transfer_spec = send_result['transfer_specs'].first['transfer_spec']
  # Add list of files to upload
  config.add_sources(transfer_spec, 'paths')
  # Start transfer
  transfer_client.start_transfer_and_wait(transfer_spec)
ensure
  transfer_client.shutdown
end
